ibis/common/patterns.py

```diff
@@ -275,6 +294,160 @@ def __rich_repr__(self):
             yield name, getattr(self, name)
 
 
+class Creator(Hashable):
+    __slots__ = ()
+
+    @abstractmethod
+    def __eq__(self, other):
+        ...
+
+    @abstractmethod
+    def make(self, context):
+        ...
+
+
+def _create_replacer(obj):
+    # TODO(kszucs): the replacer object must be handled differently from patterns
+    # basically a replacer is just a lazy way to construct objects from the context
+    # we should have a separate base class for replacers like Variable, Function,
+    # Just, Apply and Call. Something like Replacer with a specific method e.g.
+    # apply() could work
+    if isinstance(obj, Creator):
+        return obj
+    elif callable(obj):
+        # not function but something else
+        return Factory(obj)
+    else:
+        return Just(obj)
+
+
+class Variable(_Slotted, Creator):
+    """Retrieve a value from the context.
+
+    Parameters
+    ----------
+    name
+        The key to retrieve from the state.
+    """
+
+    __slots__ = ("name",)
+
+    def make(self, context):
+        return context[self]
+
+    def __getattr__(self, name):
+        return Call(operator.attrgetter(name), self)
+
+    def __getitem__(self, name):
+        return Call(operator.itemgetter(name), self)
+
+
+class Just(_Slotted, Creator):
```

Comment:
The objects here need docstrings describing what concept they are modeling.

Reply:
Adding.